File: frontend/src/api/compliance.ts
import apiClient from './client'

export interface MetaOption { value: string; label: string }

export interface ComplianceMeta {
  authorities: MetaOption[]
  frequencies: MetaOption[]
  statuses: MetaOption[]
  results: MetaOption[]
}

export interface ComplianceProgram {
  id: number
  facility_id: number
  code: string
  name: string
  description: string | null
  authority: string
  citation: string | null
  frequency: string
  discipline_id: number | null
  applies_to_space_uses: string[] | null
  applies_to_equipment_ids: number[] | null
  grace_days: number
  requires_certificate: boolean
  certificate_must_be_posted: boolean
  requires_licensed_provider: boolean
  procedure: string | null
  reading_point_codes: string[] | null
  is_active: boolean
  open_tasks: number
  overdue_tasks: number
  subject_count: number
}

export interface ComplianceTask {
  id: number
  facility_id: number
  program_id: number
  equipment_id: number | null
  location_id: number | null
  status: string
  due_date: string
  grace_days: number
  completed_at: string | null
  result: string | null
  findings: string | null
  corrective_action: string | null
  certificate_number: string | null
  certificate_issued_by: string | null
  inspector_license: string | null
  certificate_issued_on: string | null
  certificate_expires_on: string | null
  notes: string | null
  program_code: string | null
  program_name: string | null
  authority: string | null
  citation: string | null
  procedure: string | null
  requires_certificate: boolean
  requires_licensed_provider: boolean
  equipment_tag: string | null
  location_code: string | null
  is_overdue: boolean
  days_overdue: number
  // Late but inside tolerance is recoverable; past grace is a gap in the
  // record that cannot be filled retroactively.
  is_past_grace: boolean
}

export interface ComplianceSummary {
  open: number
  overdue: number
  past_grace: number
  due_within_14_days: number
  expired_certificates: number
  failures_last_12_months: number
}

export const fetchComplianceMeta = async (): Promise<ComplianceMeta> => {
  const res = await apiClient.get('/compliance/meta')
  return res.data
}

export const fetchComplianceSummary = async (
  facilityId?: number,
): Promise<ComplianceSummary> => {
  const res = await apiClient.get('/compliance/summary', {
    params: facilityId ? { facility_id: facilityId } : undefined,
  })
  return res.data
}

export const fetchPrograms = async (params: {
  facility_id?: number
  authority?: string
  discipline_id?: number
}): Promise<{ items: ComplianceProgram[]; total: number }> => {
  const res = await apiClient.get('/compliance/programs', { params })
  return res.data
}

export const seedPrograms = async (
  facilityId: number,
): Promise<{ created: number; skipped: number; programs: string[] }> => {
  const res = await apiClient.post('/compliance/programs/seed', null, {
    params: { facility_id: facilityId },
  })
  return res.data
}

export const updateProgram = async (
  id: number, payload: Partial<ComplianceProgram>,
): Promise<ComplianceProgram> => {
  const res = await apiClient.put(`/compliance/programs/${id}`, payload)
  return res.data
}

/** Idempotent — safe to re-run, and meant for a nightly job. */
export const generateTasks = async (params: {
  facility_id?: number
  horizon_days?: number
}): Promise<{ created: number; skipped_existing: number; by_program: Record<string, number> }> => {
  const res = await apiClient.post('/compliance/generate', null, { params })
  return res.data
}

export const fetchTasks = async (params: {
  facility_id?: number
  program_id?: number
  equipment_id?: number
  status?: string
  overdue_only?: boolean
  due_within_days?: number
  limit?: number
}): Promise<{ items: ComplianceTask[]; total: number }> => {
  const res = await apiClient.get('/compliance/tasks', { params })
  return res.data
}

export const completeTask = async (id: number, payload: {
  result: string
  findings?: string
  corrective_action?: string
  performed_by_vendor_id?: number
  certificate?: {
    certificate_number?: string
    certificate_issued_by?: string
    inspector_license?: string
    certificate_issued_on?: string
    certificate_expires_on?: string
  }
  measured_values?: Record<string, unknown>
  notes?: string
}): Promise<ComplianceTask> => {
  const res = await apiClient.post(`/compliance/tasks/${id}/complete`, payload)
  return res.data
}

/**
 * Distinct from the overdue task list: a posted elevator certificate can expire
 * while the next inspection is already booked, and the car still comes out of
 * service on the date.
 */
export const fetchExpiringCertificates = async (params: {
  facility_id?: number
  horizon_days?: number
}): Promise<{ items: ComplianceTask[]; total: number }> => {
  const res = await apiClient.get('/compliance/certificates/expiring', { params })
  return res.data
}

File: backend/app/schemas/compliance.py
from typing import Any, Dict, List, Optional
from datetime import date, datetime

from pydantic import BaseModel


class ComplianceProgramBase(BaseModel):
    facility_id: int
    code: str
    name: str
    description: Optional[str] = None
    authority: str
    citation: Optional[str] = None
    frequency: str
    discipline_id: Optional[int] = None
    applies_to_space_uses: Optional[List[str]] = None
    applies_to_equipment_ids: Optional[List[int]] = None
    grace_days: int = 0
    requires_certificate: bool = False
    certificate_must_be_posted: bool = False
    requires_licensed_provider: bool = False
    procedure: Optional[str] = None
    reading_point_codes: Optional[List[str]] = None


class ComplianceProgramCreate(ComplianceProgramBase):
    pass


class ComplianceProgramUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    authority: Optional[str] = None
    citation: Optional[str] = None
    frequency: Optional[str] = None
    discipline_id: Optional[int] = None
    applies_to_space_uses: Optional[List[str]] = None
    applies_to_equipment_ids: Optional[List[int]] = None
    grace_days: Optional[int] = None
    requires_certificate: Optional[bool] = None
    certificate_must_be_posted: Optional[bool] = None
    requires_licensed_provider: Optional[bool] = None
    procedure: Optional[str] = None
    reading_point_codes: Optional[List[str]] = None
    is_active: Optional[bool] = None


class ComplianceProgram(ComplianceProgramBase):
    id: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ComplianceProgramWithCounts(ComplianceProgram):
    open_tasks: int = 0
    overdue_tasks: int = 0
    subject_count: int = 0


class ComplianceProgramListResponse(BaseModel):
    items: List[ComplianceProgramWithCounts]
    total: int


class CertificateInput(BaseModel):
    certificate_number: Optional[str] = None
    certificate_issued_by: Optional[str] = None
    # A surveyor asks who signed and whether they were licensed on the day.
    inspector_license: Optional[str] = None
    certificate_issued_on: Optional[date] = None
    certificate_expires_on: Optional[date] = None


class ComplianceTaskComplete(BaseModel):
    result: str
    findings: Optional[str] = None
    corrective_action: Optional[str] = None
    performed_by_vendor_id: Optional[int] = None
    certificate: Optional[CertificateInput] = None
    measured_values: Optional[Dict[str, Any]] = None
    notes: Optional[str] = None


class ComplianceTask(BaseModel):
    id: int
    facility_id: int
    program_id: int
    equipment_id: Optional[int] = None
    location_id: Optional[int] = None
    status: str
    due_date: date
    grace_days: int
    completed_at: Optional[datetime] = None
    completed_by_id: Optional[int] = None
    performed_by_vendor_id: Optional[int] = None
    result: Optional[str] = None
    findings: Optional[str] = None
    corrective_action: Optional[str] = None
    work_order_id: Optional[int] = None
    certificate_number: Optional[str] = None
    certificate_issued_by: Optional[str] = None
    inspector_license: Optional[str] = None
    certificate_issued_on: Optional[date] = None
    certificate_expires_on: Optional[date] = None
    measured_values: Optional[Dict[str, Any]] = None
    notes: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class ComplianceTaskWithContext(ComplianceTask):
    program_code: Optional[str] = None
    program_name: Optional[str] = None
    authority: Optional[str] = None
    citation: Optional[str] = None
    procedure: Optional[str] = None
    requires_certificate: bool = False
    requires_licensed_provider: bool = False
    equipment_tag: Optional[str] = None
    location_code: Optional[str] = None
    is_overdue: bool = False
    days_overdue: int = 0
    # Late but inside tolerance is recoverable; past grace is a gap in the
    # record that cannot be filled retroactively.
    is_past_grace: bool = False


class ComplianceTaskListResponse(BaseModel):
    items: List[ComplianceTaskWithContext]
    total: int


class ComplianceSummary(BaseModel):
    open: int
    overdue: int
    past_grace: int
    due_within_14_days: int
    expired_certificates: int
    failures_last_12_months: int


class GenerationResult(BaseModel):
    created: int
    skipped_existing: int
    by_program: Dict[str, int]
    horizon_days: int


class SeedResult(BaseModel):
    created: int
    skipped: int
    programs: List[str]


class ComplianceMeta(BaseModel):
    authorities: List[Dict[str, str]]
    frequencies: List[Dict[str, str]]
    statuses: List[Dict[str, str]]
    results: List[Dict[str, str]]
